pennylane/_qubit_device.py

```diff
@@ -688,3 +697,97 @@ def sample(self, observable):
         unraveled_indices = [2] * len(device_wires)
         indices = np.ravel_multi_index(samples.T, unraveled_indices)
         return observable.eigvals[indices]
+
+    def adjoint_jacobian(self, tape):
+        """Implements the adjoint method outlined in
+        `Jones and Gacon <https://arxiv.org/abs/2009.02823>`__ to differentiate an input tape.
+
+        After a forward pass, the circuit is reversed by iteratively applying the inverse (adjoint)
+        gate to scan backwards through the circuit. This method is similar to the reversible method,
+        but has a lower time overhead and a similar memory overhead.
+
+        .. note::
+            The adjoint differentation method has the following restrictions:
+
+            * As it requires knowledge of the statevector, only statevector simulator devices can be
+              used.
+
+            * Only expectation values are supported as measurements.
+
+        Args:
+            tape (.QuantumTape): circuit that the function takes the gradient of
+
+        Returns:
+            np.ndarray: the derivative of the tape with respect to trainable parameters.
+            Dimensions are ``(len(observables), len(trainable_params))``.
+
+        Raises:
+            QuantumFunctionError: if the input tape has measurements that are not expectation values
+                or contains a multi-parameter operation aside from :class:`~.Rot`
+        """
+
+        for m in tape.measurements:
+            if m.return_type is not qml.operation.Expectation:
+                raise qml.QuantumFunctionError(
+                    "Adjoint differentiation method does not support"
+                    f" measurement {m.return_type.value}"
+                )
+
+            if not hasattr(m.obs, "base_name"):
+                m.obs.base_name = None  # This is needed for when the observable is a tensor product
+
+        # Perform the forward pass
+        self.reset()
+
+        # Consider using caching and calling lower-level functionality. We just need the state
+        # without postprocessing https://github.com/PennyLaneAI/pennylane/pull/1032/files#r563441040
+        self.execute(tape)
+
+        phi = self._reshape(self.state, [2] * self.num_wires)
+
+        lambdas = [self._apply_operation(phi, obs) for obs in tape.observables]
+
+        jac = np.zeros((len(tape.observables), len(tape.trainable_params)))
+
+        expanded_ops = []
+        for op in reversed(tape.operations):
+            if op.num_params > 1:
+                if isinstance(op, qml.Rot) and not op.inverse:
```

Comment:
It's the only multi-param op supported for now. Ones like `U2` and `CRot` are not allowed. Long term it would be great to support all!

Also we don't support `Rot` if it is inverse (I don't have a good reason, but the results were weird and I felt better to exclude it for now).

Reply:
> Also we don't support Rot if it is inverse (I don't have a good reason, but the results were weird and I felt better to exclude it for now).

Hmm.. could it be that the inverse info is lost when decomposing `qml.Rot`? Seems like the returned decomposition isn't actually inversed. :thinking: I'm actually not sure if this is intended or a bug, but could it in that case perhaps be fixed by adding something like:
```python
if op.inverse:
    for op_ in ops:
        op_.inv()
```

Reply:
That's a good point! Thanks, I made a note that inverses probably need some improvement!